--- a/dlls/dinput/mouse.c
+++ b/dlls/dinput/mouse.c
@@ -393,6 +393,19 @@
         }
     }
 
+    if (This->m_state.rgbButtons[1] == 0x80) {
+      if (!This->clipped) {
+	  RECT rect;
+	  GetClientRect(This->base.win, &rect);
+	  MapWindowPoints( This->base.win, 0, (POINT *)&rect, 2 );
+	  ClipCursor(&rect);
+	  This->clipped = TRUE;
+	}
+    } else if (This->m_state.rgbButtons[1] == 0) {
+      ClipCursor(NULL);
+      This->clipped = FALSE;
+    }
+    
     LeaveCriticalSection( &This->base.crit );
 }
 
