diff --git a/dlls/dinput/mouse.c b/dlls/dinput/mouse.c
index 7133359351..e35c6c3561 100644
--- a/dlls/dinput/mouse.c
+++ b/dlls/dinput/mouse.c
@@ -350,8 +350,22 @@ static int dinput_mouse_hook( LPDIRECTINPUTDEVICE8A iface, WPARAM wparam, LPARAM
 
             if (pt.x || pt.y)
             {
-                if ((This->warp_override == WARP_FORCE_ON) ||
-                    (This->warp_override != WARP_DISABLE && (This->base.dwCoopLevel & DISCL_EXCLUSIVE)))
+                RECT rect;
+                BOOL edge = FALSE;
+                if (GetClientRect(This->base.win, &rect))
+                {
+		    int nx = hook->pt.x + pt.x;
+		    int ny = hook->pt.y + pt.y;
+		    int b = 40;
+                    MapWindowPoints( This->base.win, 0, (POINT *)&rect, 2 );
+                    edge = nx <= rect.left + b
+		      || ny <= rect.top + b
+		      || nx >= rect.right - b
+		      || ny >= rect.bottom - b;
+                }
+                if ((This->warp_override == WARP_FORCE_ON)
+                    || (edge && This->m_state.rgbButtons[1] == 0x80)
+                    || (This->warp_override != WARP_DISABLE && (This->base.dwCoopLevel & DISCL_EXCLUSIVE)))
                     This->need_warp = TRUE;
             }
             break;
