diff --git a/dlls/dinput/mouse.c b/dlls/dinput/mouse.c
index 03fc0b025e..5806129e81 100644
--- a/dlls/dinput/mouse.c
+++ b/dlls/dinput/mouse.c
@@ -358,8 +358,17 @@ static int dinput_mouse_hook( LPDIRECTINPUTDEVICE8A iface, WPARAM wparam, LPARAM
 
             if (pt.x || pt.y)
             {
-                if ((This->warp_override == WARP_FORCE_ON) ||
-                    (This->warp_override != WARP_DISABLE && (This->base.dwCoopLevel & DISCL_EXCLUSIVE)))
+                RECT rect;
+                BOOL edge = FALSE;
+                if (GetClientRect(This->base.win, &rect))
+                {
+                    MapWindowPoints( This->base.win, 0, (POINT *)&rect, 2 );
+                    edge = hook->pt.x < 2 || hook->pt.y < 2 || hook->pt.x > ((rect.left + rect.right) - 2)
+                        || hook->pt.y > ((rect.top + rect.bottom) - 2);
+                }
+                if ((This->warp_override == WARP_FORCE_ON)
+                    || (edge && This->m_state.rgbButtons[1] == 0x80)
+                    || (This->warp_override != WARP_DISABLE && (This->base.dwCoopLevel & DISCL_EXCLUSIVE)))
                     This->need_warp = TRUE;
             }
             break;
